File: projects/samples/devices/controllers/connector/connector.py
# Copyright 1996-2023 Cyberbotics Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
An example of controller using a Connector device.
"""

from controller import Robot
import math


class Controller(Robot):
    SPEED = 2
    timeStep = 64

    def __init__(self):
        super(Controller, self).__init__()
        self.connector = self.getDevice('connector')
        self.connector.enablePresence(self.timeStep)
        self.upper_motor = self.getDevice('upper motor')
        self.upper_motor.setPosition(-1.57)
        self.upper_position_sensor = self.getDevice('upper sensor')
        self.upper_position_sensor.enable(self.timeStep)
        self.left_motor = self.getDevice('left wheel motor')
        self.right_motor = self.getDevice('right wheel motor')
        self.left_motor.setPosition(float('inf'))
        self.right_motor.setPosition(float('inf'))
        self.left_motor.setVelocity(0.0)
        self.right_motor.setVelocity(0.0)

    def run(self):
        robot_number = int(self.getName()[6])
        hop_counts = 0
        direction = 1
        state = 'CONNECTOR_IN_PLACE'
        while self.step(self.timeStep) != -1:
            left_speed = 0
            right_speed = 0
            if state == 'CONNECTOR_IN_PLACE':  # wait for the motor to be in place
                if math.fabs(self.upper_position_sensor.getValue() + (direction * 1.56)) < 0.01:
                    state = 'CONNECTING'
            elif state == 'CONNECTING':  # wait for another Connector device to lock with
                if self.connector.getPresence() == 1:
                    self.connector.lock()
                    self.upper_motor.setPosition(direction * 1.57)
                    state = 'PASSING_OVER'
                # if it is the robot which is moving, we need to take it closer to
                # the other one so that they can attach to each other
                if robot_number == 1:
                    left_speed = direction * self.SPEED
                    right_speed = direction * self.SPEED
            elif state == 'PASSING_OVER':  # wait for the motor to be in the new place (hopping)
                if math.fabs(self.upper_position_sensor.getValue() - (direction * 1.56)) < 0.01:
                    self.connector.unlock()
                    state = 'NEXT_HOP'
                    hop_counts += 1
            elif state == 'NEXT_HOP':  # wait for the other Connector device to be far enough
                if self.connector.getPresence() == 0:
                    # if it is the robot which is moving, we need to replace the
                    # motor in order to be able to atach to the next robot
                    if robot_number == 1:
                        if hop_counts % 2 == 0:
                            direction *= -1
                        self.upper_motor.setPosition(-1.57 * direction)
                        state = 'CONNECTOR_IN_PLACE'

                    # If not, the motor is already in the right place as the moving
                    # robot will come back at the same place.
                    else:
                        state = 'CONNECTING'
                        direction *= -1
                # the moving robot needs to get away from the connector in order to
                # avoid attaching always to the same robot.
                if robot_number == 1:
                    left_speed = direction * self.SPEED
                    right_speed = direction * self.SPEED
            # set the motor speeds
            self.left_motor.setVelocity(left_speed)
            self.right_motor.setVelocity(right_speed)


controller = Controller()
controller.run()

File: projects/samples/devices/controllers/connector/connector.c
/*
 * Copyright 1996-2023 Cyberbotics Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * Description:  An example of controller using a Connector device.
 */

#include <math.h>
#include <stdlib.h>
#include <webots/connector.h>
#include <webots/motor.h>
#include <webots/position_sensor.h>
#include <webots/robot.h>

#define SPEED 2
#define TIME_STEP 64

typedef enum { CONNECTOR_IN_PLACE, CONNECTING, PASSING_OVER, NEXT_HOP } states;

int main() {
  WbDeviceTag connector, upper_motor, upper_position_sensor, left_motor, right_motor;
  /*
   * we use these different variables for:
   *   - robot_number : To differentiate the robots
   *   - hop_counts : To know when to invert the direction
   *   - direction : To know in which direction we will perform the next hop
   */
  int robot_number, hop_counts = 0, direction = 1;
  states state;

  wb_robot_init();

  /* get a handler to the connector, the motor and the position_sensor. */
  connector = wb_robot_get_device("connector");
  upper_motor = wb_robot_get_device("upper motor");
  upper_position_sensor = wb_robot_get_device("upper sensor");

  /* activate them. */
  wb_connector_enable_presence(connector, TIME_STEP);
  wb_position_sensor_enable(upper_position_sensor, TIME_STEP);

  wb_motor_set_position(upper_motor, -1.57);

  /* get a handler to the motors and set target position to infinity (speed control). */
  left_motor = wb_robot_get_device("left wheel motor");
  right_motor = wb_robot_get_device("right wheel motor");
  wb_motor_set_position(left_motor, INFINITY);
  wb_motor_set_position(right_motor, INFINITY);
  wb_motor_set_velocity(left_motor, 0.0);
  wb_motor_set_velocity(right_motor, 0.0);

  /*
   * we need to substract the ASCII code of '0' (i.e., 48) to convert
   * the char of the number into an integer.
   */
  robot_number = (wb_robot_get_name())[6] - '0';
  state = CONNECTOR_IN_PLACE;

  while (wb_robot_step(TIME_STEP) != -1) {
    int left_speed = 0;
    int right_speed = 0;

    switch (state) {
      /* wait for the motor to be in place. */
      case CONNECTOR_IN_PLACE:
        if (fabs(wb_position_sensor_get_value(upper_position_sensor) + (direction * 1.56)) < 0.01)
          state = CONNECTING;
        break;

      /* wait for another Connector device to lock with. */
      case CONNECTING:
        if (wb_connector_get_presence(connector) == 1) {
          wb_connector_lock(connector);
          wb_motor_set_position(upper_motor, direction * 1.57);
          state = PASSING_OVER;
        }

        /*
         * if it is the robot which is moving, we need to take it closer to
         * the other one so that they can attach to each other
         */
        if (robot_number == 1) {
          left_speed = direction * SPEED;
          right_speed = direction * SPEED;
        }
        break;

      /* wait for the motor to be in the new place (hopping). */
      case PASSING_OVER:
        if (fabs(wb_position_sensor_get_value(upper_position_sensor) - (direction * 1.56)) < 0.01) {
          wb_connector_unlock(connector);
          state = NEXT_HOP;
          hop_counts++;
        }
        break;

      /* wait for the other Connector device to be far enough. */
      case NEXT_HOP:
        if (wb_connector_get_presence(connector) == 0) {
          /*
           * if it is the robot which is moving, we need to replace the
           * motor in order to be able to atach to the next robot
           */
          if (robot_number == 1) {
            if (hop_counts % 2 == 0)
              direction *= -1;
            wb_motor_set_position(upper_motor, -1.57 * direction);
            state = CONNECTOR_IN_PLACE;

            /*
             * If not, the motor is already in the right place as the moving
             * robot will come back at the same place.
             */
          } else {
            state = CONNECTING;
            direction *= -1;
          }
        }

        /*
         * the moving robot needs to get away from the connector in order to
         * avoid attaching always to the same robot.
         */
        if (robot_number == 1) {
          left_speed = direction * SPEED;
          right_speed = direction * SPEED;
        }
        break;
    }
    /* set the motor speeds. */
    wb_motor_set_velocity(left_motor, left_speed);
    wb_motor_set_velocity(right_motor, right_speed);
  }

  wb_robot_cleanup();

  return 0;
}
